executor/aggregate.go

```diff
@@ -314,9 +316,18 @@ func (w *HashAggPartialWorker) getChildInput() bool {
 	return true
 }
 
+func recoveryHashAgg(output chan *AfFinalResult, r interface{}) {
+	output <- &AfFinalResult{err: errors.Errorf("%v", r)}
+	buf := util.GetStack()
+	log.Errorf("panic in the recoverable goroutine: %v, stack trace:\n%s", r, buf)
+}
+
 func (w *HashAggPartialWorker) run(ctx sessionctx.Context, waitGroup *sync.WaitGroup, finalConcurrency int) {
```

Comment:
It's better to use `util.WithRecovery()`

Reply:
I tried, but for this two operator using `util.WithRecovery()` is not suitable, except do recovery, we need take param `waitGroup` or even local variable `output` or `needShuffle` to do some cleanup operation(and recovery send response must be done before those cleanup), use withRecovery's closure to catch those variables will make more complex code, I think make reuseable `recoveryXXX` to recovery and send result is more simple and clear to this snippet.

Reply:
Can we put the cleanup into a `recoverFn` of `WithRecovery`? like `HashJoinExec::finishOuterFetcher`?

Reply:
@eurekaka `HashJoinExec::finishOuterFetcher` only depend on `HashJoinExec`'s fields, but here some local vairable like `needShuffle` or `projectionOutput` is local variable, we need modify more to share them to `recoveFn`.